feat: engine assets and proxy design for guardrails

themaherkhalil opened the pull request:
Base commit for all the interceptors and guardrails. EngineProxyFactory - Takes the real engine and generates a proxy from it PipelineInvocationHandler - Uses the provided JSON in the engine SMSS instrumented through PIPELINE to get output from it IInputReactor, IOutputReactor - Marker interface for input interception and output interception PHI, PII, Logging - Sample Reactor
Utility - methods to take the engine go to engineproxy and get the proxied engine IEngine - the PIPELINE constant
Sample JSON
{
  "pipelines": {
    "ask": {
      "input": [
        {
          "reactorClass": "prerna.reactor.interceptor.LoggingInputReactor",
          "params": {
            "logLevel": "INFO",
            "logMessage": "Executing ASK method..."
          }
        },
        {
          "reactorClass": "prerna.reactor.interceptor.PIIInputReactor",
          "params": {
            "blockOnPII": true,
            "piiTypes": ["SSN", "CREDIT_CARD"]
          }
        }
      ],
      "output": [
        {
          "reactorClass": "prerna.reactor.interceptor.PHIOutputReactor",
          "params": {
            "redactionMask": "[REDACTED_PHI]"
          }
        }
      ]
    }
  }
}

prabhuk12 committed:
feat: Interceptors and Guardrails

Base commit for all the interceptors and guardrails.
EngineProxyFactory - Takes the real engine and generates a proxy from it
PipelineInvocationHandler - Uses the provided JSON in the engine SMSS instrumented through PIPELINE to get output from it
IInputReactor, IOutputReactor - Marker interface for input interception and output interception
PHI, PII, Logging - Sample Reactor
Utility - methods to take the engine go to engineproxy and get the proxied engine
IEngine - the PIPELINE constant
Sample JSON
{
  "pipelines": {
    "ask": {
      "input": [
        {
          "reactorClass": "prerna.reactor.interceptor.LoggingInputReactor",
          "params": {
            "logLevel": "INFO",
            "logMessage": "Executing ASK method..."
          }
        },
        {
          "reactorClass": "prerna.reactor.interceptor.PIIInputReactor",
          "params": {
            "blockOnPII": true,
            "piiTypes": ["SSN", "CREDIT_CARD"]
          }
        }
      ],
      "output": [
        {
          "reactorClass": "prerna.reactor.interceptor.PHIOutputReactor",
          "params": {
            "redactionMask": "[REDACTED_PHI]"
          }
        }
      ]
    }
  }
}

CodiumAI-Agent commented:
## Title

feat: interceptors and guardrails

___
### **User description**
Base commit for all the interceptors and guardrails. EngineProxyFactory - Takes the real engine and generates a proxy from it PipelineInvocationHandler - Uses the provided JSON in the engine SMSS instrumented through PIPELINE to get output from it IInputReactor, IOutputReactor - Marker interface for input interception and output interception PHI, PII, Logging - Sample Reactor
Utility - methods to take the engine go to engineproxy and get the proxied engine IEngine - the PIPELINE constant
Sample JSON
{
  "pipelines": {
    "ask": {
      "input": [
        {
          "reactorClass": "prerna.reactor.interceptor.LoggingInputReactor",
          "params": {
            "logLevel": "INFO",
            "logMessage": "Executing ASK method..."
          }
        },
        {
          "reactorClass": "prerna.reactor.interceptor.PIIInputReactor",
          "params": {
            "blockOnPII": true,
            "piiTypes": ["SSN", "CREDIT_CARD"]
          }
        }
      ],
      "output": [
        {
          "reactorClass": "prerna.reactor.interceptor.PHIOutputReactor",
          "params": {
            "redactionMask": "[REDACTED_PHI]"
          }
        }
      ]
    }
  }
}


___

### **PR Type**
Enhancement


___

### **Description**
- EngineProxyFactory adds dynamic engine proxy wrapping

- PipelineInvocationHandler handles input/output reactor pipelines

- Add Logging, PII, PHI reactor implementations

- Wrap engine getters in Utility with proxies


___

### **Changes diagram**

```mermaid
flowchart LR
  RealEngine["Real Engine"]
  Proxy["Engine Proxy"]
  Handler["PipelineInvocationHandler"]
  Input["Input Reactors"]
  Method["Real Engine Method"]
  Output["Output Reactors"]
  Result["Final Result"]

  RealEngine -- "wrap by EngineProxyFactory" --> Proxy
  Proxy -- "invokes" --> Handler
  Handler -- "run input pipelines" --> Input
  Input -- "invoke real engine" --> Method
  Method -- "run output pipelines" --> Output
  Output -- "produce final result" --> Result
```


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>IEngine.java</strong><dd><code>Add PIPELINE constant to IEngine</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-4d297d94829c703c2e550774225d256d6834f33051f386c113086a1221f59821">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td><strong>EngineProxyFactory.java</strong><dd><code>Add engine proxy factory methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-68ea54768ed9b831ce5e30d02baf04d83c849249c3f9361c7055cf0a4facfa81">+131/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>PipelineInvocationHandler.java</strong><dd><code>Implement pipeline invocation handler logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-cb730db4c8d4106ac4fb58cebeab9a62ac0baef97764b2ff594f48dab4bc3810">+266/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>IInputReactor.java</strong><dd><code>Add IInputReactor marker interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-092086572b8c4b277f5cbff971245625a677cc573ca3f0c2027a7bc9e92bac98">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>IOutputReactor.java</strong><dd><code>Add IOutputReactor marker interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-1ad697701d9d86c270ea9ffcce4a7b2973ff7c696526f091373d441f335acbee">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>LoggingInputReactor.java</strong><dd><code>Implement logging input reactor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-0ba83d6b4f8e7037a4bfcc7263f1c4302be8845f867299807d9d5466dad799a7">+58/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>PIIInputReactor.java</strong><dd><code>Implement PII input guardrail reactor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-e2c02addf661e13f582c692598e7e2deb4bc102d77b7767e9a30ddfdbe63c6f8">+72/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>PHIOutputReactor.java</strong><dd><code>Implement PHI output redaction reactor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-3190f4d60e5025c55a97097193c3723e77eded2cb3c99abc356345f657cc8569">+57/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>PipelineReactorUtils.java</strong><dd><code>Add pipeline utility constant keys</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-804c6252512b8ec0758afd2a432aa8b3d8081b69bbf64b6d67437e0cf9217eb6">+23/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Utility.java</strong><dd><code>Wrap engine getters with proxy calls</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1044/files#diff-baa3c1b007e2168b58afab6a8cf28a288a718bafb5441f4331f9e19f110a6b0d">+30/-5</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

CodiumAI-Agent commented:
## PR Code Suggestions ✨

<!-- c00d2f1 -->

<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=4>Possible issue</td>
<td>



<details><summary>Return transformed result</summary>

___


**Return the processed result from the output pipeline (<code>processedResult</code>) instead of <br>the original <code>result</code> to ensure that output interceptors' modifications are propagated <br>to the caller. Otherwise, any redactions or transformations are ignored.**

[src/prerna/engine/impl/pipeline/PipelineInvocationHandler.java [126]](https://github.com/SEMOSS/Semoss/pull/1044/files#diff-cb730db4c8d4106ac4fb58cebeab9a62ac0baef97764b2ff594f48dab4bc3810R126-R126)

```diff
-return result;
+return processedResult;
```
<details><summary>Suggestion importance[1-10]: 9</summary>

__

Why: Returning `result` ignores modifications from output reactors, so switching to `processedResult` ensures that any transformations or redactions are applied to the caller.


</details></details></td><td align=center>High

</td></tr><tr><td>



<details><summary>Set default PASS to true</summary>

___


**Use <code>PASS = true</code> by default when no PII is detected so that non-blocking inputs <br>proceed correctly. Only set it to false when a PII violation forces a block.**

[src/prerna/reactor/interceptor/PIIInputReactor.java [68]](https://github.com/SEMOSS/Semoss/pull/1044/files#diff-e2c02addf661e13f582c692598e7e2deb4bc102d77b7767e9a30ddfdbe63c6f8R68-R68)

```diff
-resultMap.put(PipelineReactorUtils.PASS, false);
+resultMap.put(PipelineReactorUtils.PASS, true);
```
<details><summary>Suggestion importance[1-10]: 9</summary>

__

Why: Setting `PASS` to false unconditionally blocks all `ask` calls; defaulting to `true` only blocks when PII is detected fixes the logic so non-PII inputs proceed correctly.


</details></details></td><td align=center>High

</td></tr><tr><td>



<details><summary>Initialize argument mapping early</summary>

___


**Initialize <code>processedArguments</code> to a map of original method arguments before running <br>the input pipeline to avoid NullPointerExceptions and ensure correct argument <br>mapping. Populate this map using the method parameter names and incoming <code>args</code>. This <br>provides a baseline for pipeline reactors when no modifications occur.**

[src/prerna/engine/impl/pipeline/PipelineInvocationHandler.java [63]](https://github.com/SEMOSS/Semoss/pull/1044/files#diff-cb730db4c8d4106ac4fb58cebeab9a62ac0baef97764b2ff594f48dab4bc3810R63-R63)

```diff
-Map<String, Object> processedArguments = null;
+Map<String, Object> processedArguments = new HashMap<>();
+Parameter[] parameters = method.getParameters();
+if (args != null) {
+    for (int i = 0; i < parameters.length; i++) {
+        processedArguments.put(parameters[i].getName(), args[i]);
+    }
+}
```
<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: Without setting `processedArguments` before the pipelines, `unmapArguments` may receive a null map causing NPEs; this change ensures a valid map of original arguments is always available.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Guard pipeline parsing on null</summary>

___


**Guard against a null or empty pipeline JSON before parsing to avoid runtime <br>exceptions when no configuration file is found. Skip pipeline loading when <br><code>pipelineJson</code> is null or empty.**

[src/prerna/engine/impl/pipeline/PipelineInvocationHandler.java [44-45]](https://github.com/SEMOSS/Semoss/pull/1044/files#diff-cb730db4c8d4106ac4fb58cebeab9a62ac0baef97764b2ff594f48dab4bc3810R44-R45)

```diff
 String pipelineJson = getJsonData(realEngine);
-parseAndLoadPipelines(pipelineJson);
+if (pipelineJson != null && !pipelineJson.isEmpty()) {
+    parseAndLoadPipelines(pipelineJson);
+}
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: If `getJsonData` returns null or empty, `parseAndLoadPipelines` will throw exceptions; checking for null/empty protects against missing or invalid pipeline configs.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

prabhuk12 committed:
feat: Interceptors

Quick update to keep the result as part of the arguments so the reactors can mutate it if need be.

prabhuk12 committed:
feat: Interceptors

Adding check for model

themaherkhalil committed:
chore: formatting code changes

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
feat: also accounting for IRDBMSEngine interface

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
feat: build out of engine assets folder and moving files dynamically into new directory

themaherkhalil committed:
fix: using proper util to pull from any engine assets folder if the pipeline exists

themaherkhalil committed:
Merge branch 'dev' into interceptors3

themaherkhalil committed:
fix: ignoring push to cloud for default engines, smss concealing for SERVICE_ACCOUNT_CREDENTIALS

themaherkhalil committed:
fix: making sure the parent dir also exists before trying to write the owl file

themaherkhalil committed:
fix: returning the original prop instead of null

themaherkhalil committed:
fix: using dev image and removing vmounts w/ minio

themaherkhalil committed:
fix: extending isBasic to IEngine so assets directory not created for insights db or owl db
